### PR TITLE
compatible with slime metric reconfiguration

### PR DESCRIPTION
**Pull request by cywang1905**
related to https://github.com/slime-io/slime/pull/110

slime-metric-v2-architecture

### CODE REVIEW

File: module/module.go [hunk 1]
```diff
@@ -23,18 +23,41 @@ import (
...
+	mo.env = env
+}
+
+func (mo *Module) RestConfig() rest.Config {
+	return mo.restConfig
 }
 
-func (m *Module) Config() proto.Message {
-	return &m.config
+func (mo *Module) SetRestConfig(cfg rest.Config) {
```
Comment (YonkaFang) [inline module/module.go:52]: RestConfig is a too-specific thing

File: go.mod [hunk 2]
```diff
@@ -21,6 +21,6 @@ replace (
 	k8s.io/api => k8s.io/api v0.17.2
 	k8s.io/apimachinery => k8s.io/apimachinery v0.17.2
 
-	slime.io/slime/framework => github.com/slime-io/slime/framework v0.3.1
-	//slime.io/slime/framework => ../slime/framework
+	//slime.io/slime/framework => github.com/slime-io/slime/framework v0.3.1
+	slime.io/slime/framework => ../slime/framework
```
Comment (YonkaFang) [inline go.mod:25]: replace back to a specific version

Comment (cywang1905) [reply go.mod:28]: will change after slime pr finished

File: controllers/servicefence_controller.go [hunk 3]
```diff
@@ -127,22 +135,94 @@ func (r *ServicefenceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+		r.interestMeta.Set(req.NamespacedName.String(), true)
 		err = r.refreshSidecar(instance)
 	}
 
 	return ctrl.Result{}, err
 }
 
+// todo take place, reconciler call back function for watcher producer
+func (r *ServicefenceReconciler) handleWatcherEvent(event trigger.WatcherEvent) map[types.NamespacedName][]string {
+	r.interestMeta.Get(event.NN.String())
```
Comment (YonkaFang) [inline controllers/servicefence_controller.go:153]: `Get` contains process-logic is weird

File: controllers/servicefence_controller.go [hunk 4]
```diff
@@ -114,8 +126,17 @@ func (r *ServicefenceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		if errors.IsNotFound(err) {
 			// TODO should be recovered? maybe we should call refreshFenceStatusOfService here
 			log.Info("serviceFence is deleted")
-			r.source.WatchRemove(req.NamespacedName)
-			return reconcile.Result{}, nil
+			//r.interestMeta.Pop(req.NamespacedName.String())
+			delete(r.interestMeta, req.NamespacedName.String())
+			for k, v := range r.interestMeta {
+				r.InterestMetaCopy[k] = v
```
Comment (YonkaFang) [inline controllers/servicefence_controller.go:132]: wrong usage

File: controllers/servicefence_controller.go [hunk 5]
```diff
@@ -127,22 +148,143 @@ func (r *ServicefenceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
 
 	// 资源更新
 	diff := r.updateVisitedHostStatus(instance)
 	r.recordVisitor(instance, diff)
 	if instance.Spec.Enable {
-		if r.source != nil {
-			r.source.WatchAdd(types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace})
+		r.interestMeta[req.NamespacedName.String()] = true
+		for k, v := range r.interestMeta {
+			r.InterestMetaCopy[k] = v
```
Comment (YonkaFang) [inline controllers/servicefence_controller.go:159]: wrong use

File: controllers/servicefence_controller.go [hunk 6]
```diff
@@ -127,22 +148,143 @@ func (r *ServicefenceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+	r.reconcileLock.RLock()
+	defer r.reconcileLock.RUnlock()
+
+	// check event
+	gvks := []schema.GroupVersionKind{
+		{Group: "networking.istio.io", Version: "v1beta1", Kind: "Sidecar"},
+	}
+	for _, gvk := range gvks {
+		if event.GVK == gvk && r.InterestMetaCopy[event.NN.String()] {
+			goto GenerateQueryMap
```
Comment (YonkaFang) [inline controllers/servicefence_controller.go:183]: goto should be very careful to use. here we have more approaches

File: controllers/servicefence_controller.go [hunk 7]
```diff
@@ -127,22 +148,143 @@ func (r *ServicefenceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+				EventChan: make(chan trigger.TickerEvent),
+			},
+			PrometheusSourceConfig: prometheusSourceConfig,
+		},
+		StopChan: env.Stop,
+	}, nil
+
+}
+
+func newPrometheusSource(env bootstrap.Environment) (metric.PrometheusSourceConfig, error) {
```
Comment (YonkaFang) [inline controllers/servicefence_controller.go:268]: newsource -> newsourceconfig?

File: controllers/servicefence_controller.go [hunk 8]
```diff
@@ -127,22 +148,143 @@ func (r *ServicefenceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+		hs = append(hs, h)
+	}
+	qm[event.NN.String()] = hs
+	return qm
+}
+
+// call back function for ticker producer
+func (r *ServicefenceReconciler) handleTickerEvent(event trigger.TickerEvent) metric.QueryMap {
+	r.reconcileLock.RLock()
+	defer r.reconcileLock.RUnlock()
```
Comment (YonkaFang) [inline controllers/servicefence_controller.go:205]:
```go
	r.reconcileLock.RLock()
        instMeta := r.interestMetaCopy
	r.reconcileLock.RUnlock()

      for ... {}
```

File: controllers/servicefence_controller.go [hunk 9]
```diff
@@ -127,22 +148,143 @@ func (r *ServicefenceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+		}
+	}
+	return nil
+
+	// generate query map for producer
+GenerateQueryMap:
+	qm := make(map[string][]metric.Handler)
+	var hs []metric.Handler
+	for name, pHandler := range r.env.Config.Metric.Prometheus.Handlers {
+		query := strings.ReplaceAll(pHandler.Query, "$namespace", event.NN.Namespace)
```
Comment (YonkaFang) [inline controllers/servicefence_controller.go:193]: we have two copies of this code block nearby

File: controllers/process.go [hunk 10]
```diff
@@ -27,25 +24,48 @@ const (
...
+			r.ConsumeMetric(metric)
 		}
-	}()
+	}
+
 }
 
-func (r *ServicefenceReconciler) Refresh(req reconcile.Request, args map[string]string) (reconcile.Result, error) {
+func (r *ServicefenceReconciler) ConsumeMetric(metric metric.Metric) {
+	for meta, results := range metric {
```
Comment (YonkaFang) [inline controllers/process.go:51]: better to serialize the execution

File: controllers/servicefence_controller.go (hunk 5 above)
Comment (YonkaFang) [reply controllers/servicefence_controller.go:159]:
```
m, mc map[string]string

func update() {
  // ...
  mut.Lock()
  delete(m, "key1")
  newMc := map[string]string{}
  for k, v := range m {
    newMc[key] = value
  }
  mc = newMc
  mut.Unlock()
}

func GetM() map[string]string {
  mut.RLock()
  return mc
  defer mut.RUnlock()
}

```

File: controllers/servicefence_controller.go (hunk 4 above)
Comment (cywang1905) [reply controllers/servicefence_controller.go]: changed according to concept of copy on write

File: controllers/servicefence_controller.go (hunk 9 above)
Comment (cywang1905) [reply controllers/servicefence_controller.go]: replace by function generateHandler

File: go.mod (hunk 2 above)
Comment (cywang1905) [reply go.mod]: change to v0.3.3

**Review by YonkaFang — APPROVED** [review]

**Pull request merged**